lib/flash.js

```diff
@@ -0,0 +1,184 @@
+}
+
+function checkChipId(chipId, next) {
+  if (Buffer.compare(chipId, CHIP_ID) !== 0) {
+    return next(new Error('Invalid chip ID (flash communication error)'));
+  }
+  next();
+}
+
+function setWriteEnabled(usbConnection, next) {
+  transaction(usbConnection, [0x06], null, null, null, next);
+}
+
+function eraseChip(usbConnection, next) {
+  transaction(usbConnection, [0x60], null, null, null, next);
+}
+
+// Poll for the WIP bit in the status register to go low
+function waitTransactionComplete(usbConnection, next) {
+  setTimeout(function onWait() {
+    getStatus(usbConnection, function onGetStatus(err, status) {
+      if (err) {
+        return next(err);
+      }
+      if (status === 0) {
+        return next();
+      }
+      process.nextTick(waitTransactionComplete, usbConnection, next);
+    });
+  }, 200);
+}
+
+// Read the status register
+function getStatus(usbConnection, next) {
+  transaction(usbConnection, [0x05], 1, null, null, function onGetStatus(err, data) {
+    next(err, data ? data[0] : data);
+  });
+}
+
+function write(usbConnection, writeAddress, data, sliceStart, next) {
+  if (!next) {
+    next = sliceStart;
+    sliceStart = 0;
+  }
+  if (sliceStart >= data.length) {
+    return next();
+  }
+
+  var sliceEnd = sliceStart + PAGE;
+  var pageData = data.slice(sliceStart, sliceEnd);
+  writePage(usbConnection, writeAddress, pageData, function onWrite(err) {
+    if (err) {
+      return next(err);
+    }
+    process.nextTick(write, usbConnection, writeAddress + PAGE, data, sliceEnd, next);
+  });
+}
+
+function concatToBuffer() {
+  var totalLength = 0;
+  var buffers = [];
+  var buffer;
+  for (var i = 0; i < arguments.length; i++) {
+    var arg = arguments[i];
+    buffer = new Buffer(arg);
+    totalLength += buffer.length;
+    buffers.push(buffer);
+  }
+  var result = new Buffer(totalLength);
+
+  var byteIndex = 0;
+  for (i = 0; i < buffers.length; i++) {
+    buffer = buffers[i];
+    buffer.copy(result, byteIndex);
+    byteIndex += buffer.length;
+  }
+
+  return result;
+}
+
+// Write a page to flash
+function writePage(usbConnection, addr, data, next) {
+  var status_poll = true;
+  var wren = true;
+  transaction(
+    usbConnection,
+    concatToBuffer([0x12], address(addr), data),
+    null,
+    status_poll,
+    wren,
+    next
+  );
+}
+
+// Returns a random integer, both low and high are inclusive
+function randint(low, high) {
+  return Math.floor(Math.random() * (high - low + 1) + low);
+}
+
+function randbyte() {
+  return randint(0, 255);
+}
+
+function asyncLog(msg, next) {
+  logs.info(msg);
+  next();
+}
+
+module.exports = function flashDevice(usbConnection, ubootBuffer, squashfsBuffer, next) {
+  var uid = [randbyte(), randbyte(), randbyte(), randbyte()];
+  var mac1 = [0x02, 0xa3].concat(uid);
+  var mac2 = [0x02, 0xa4].concat(uid);
+
+  async.waterfall([
+    asyncLog.bind(null, 'Checking the chip id...'),
+    readChipId.bind(null, usbConnection),
+    checkChipId,
```

Comment:
This information is actually how the CLI determines which USB peripheral is a Tessel, so I believe it's unnecessary, since the operation wouldn't have reached this point if the id was incorrect. Note that this value is called "TESSEL_VID" (0x1209) elsewhere.